baseplate/frameworks/queue_consumer/kombu.py

```diff
@@ -127,6 +133,148 @@ def handle(self, message: kombu.Message) -> None:
 
 
+class KombuBatchMessageHandler(MessageHandler):
+    def __init__(
+        self,
+        baseplate: Baseplate,
+        name: str,
+        handler_fn: BatchHandler,
+        error_handler_fn: Optional[BatchErrorHandler] = None,
+    ):
+        self.baseplate = baseplate
+        self.name = name
+        self.handler_fn = handler_fn
+        self.error_handler_fn = error_handler_fn
+
+    def handle(self, messages: Sequence[kombu.Message]) -> None:  # pylint: disable=arguments-differ
+        logger.info("Processing batch with %i messages", len(messages))
+
+        context = self.baseplate.make_context_object()
+        try:
+            # We place the call to ``baseplate.make_server_span`` inside the
+            # try/except block because we still want Baseplate to see and
+            # handle the error (publish it to error reporting)
+            with self.baseplate.make_server_span(context, self.name) as span:
+                span.set_tag("kind", "batch_consumer")
+
+                message: kombu.Message
+                for message in messages:
+                    delivery_info = message.delivery_info
+                    span.set_tag("amqp.routing_key", delivery_info.get("routing_key", ""))
+                    span.set_tag("amqp.consumer_tag", delivery_info.get("consumer_tag", ""))
+                    span.set_tag("amqp.delivery_tag", delivery_info.get("delivery_tag", ""))
+                    span.set_tag("amqp.exchange", delivery_info.get("exchange", ""))
+
+                self.handler_fn(context, list(messages))
+        except Exception as exc:  # pylint: disable=broad-except
+            logger.exception(
+                "Unhandled error while trying to process a message.  The message "
+                "has been returned to the queue broker."
+            )
+            if self.error_handler_fn:
+                self.error_handler_fn(context, messages, exc)
+            else:
+                for message in messages:
+                    if not message.acknowledged:
+                        message.requeue()
+
+            if isinstance(exc, FatalMessageHandlerError):
+                logger.info("Recieved a fatal error, terminating the server.")
+                raise
+        else:
+            for message in messages:
+                if not message.acknowledged:
+                    message.ack()
+            logger.info("Successfully processed batch with %i messages", len(messages))
+
+
+class KombuBatchQueueConsumer(QueueConsumer):
+    """
+    Wrapper around a MessageHandler object that interfaces with the work_queue and starts/stops the handle loop.
+
+    The batched queue consumer consumes messages until the batch_size is reached
+    or a the batch_timeout has occurred. Then, if the batch is not empty, the
+    messages are consumed. The timer is stopped after a batch is processed if
+    there are no new messages and started when there is at least one message in
+    the batch.
+    """
```

Comment:
I don't think extending the base `QueueConsumer` is really necessary, couldn't the batch gathering be done in a now pump worker and instead of putting work directly on the `work_queue` it just does what this is doing (accumulating  work into a batch and then putting the batch on the queue)?

Reply:
Makes sense. I chose to extend the QueueConsumer because I felt it would make more sense to make this the responsibility of the consumer rather than the pump worker.